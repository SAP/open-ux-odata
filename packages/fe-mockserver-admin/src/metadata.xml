<?xml version="1.0" encoding="utf-8"?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
    <edmx:DataServices>
        <Schema Namespace="MockServerAdmin" xmlns="http://docs.oasis-open.org/odata/ns/edm">
            <EntityType Name="Service">
                <Key>
                    <PropertyRef Name="UrlPath"/>
                </Key>
                <Property Name="UrlPath" Type="Edm.String" Nullable="false" MaxLength="255"/>
                <Property Name="InternalName" Type="Edm.String" Nullable="true" MaxLength="100"/>
                <Property Name="EncodedPath" Type="Edm.String" Nullable="true" MaxLength="255"/>
                <Property Name="ContextIsolation" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="Debug" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="Watch" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="HasETag" Type="Edm.Boolean" Nullable="false"/>
                <Property Name="RegisteredAt" Type="Edm.DateTimeOffset" Nullable="false"/>
            </EntityType>
            <EntityContainer Name="AdminContainer">
                <EntitySet Name="Services" EntityType="MockServerAdmin.Service"/>
            </EntityContainer>
        </Schema>
    </edmx:DataServices>
</edmx:Edmx>