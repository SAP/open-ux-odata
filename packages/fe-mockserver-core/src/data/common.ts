import type { Action, EntitySet, EntityType, Property } from '@sap-ux/vocabularies-types';
import type { ODataMetadata } from './metadata';
import type { FileBasedMockData, KeyDefinitions } from '../mockdata/fileBasedMockData';
import type { ILogger } from '@ui5/logger';
import type ODataRequest from '../request/odataRequest';
import type { IFileLoader } from '../index';

export interface EntitySetInterface {
    checkKeyValue(mockData: object, keyValues: object, keyName: string, keyProp?: Property): boolean;
    checkFilter(mockData: object, filterExpression: any, tenantId: string, odataRequest: ODataRequest): boolean;
    checkSearch(mockData: object, searchQueries: string[], odataRequest: ODataRequest): boolean;
    executeAction(
        actionDefinition: Action,
        actionData: object | undefined,
        odataRequest: ODataRequest,
        keys: Record<string, any>
    ): Promise<any>;
    performGET(
        keyValues: KeyDefinitions,
        asArray: boolean,
        tenantId: string,
        odataRequest: ODataRequest,
        dontClone?: boolean
    ): Promise<any>;
    performPOST(
        keyValues: KeyDefinitions,
        postData: any,
        tenantId: string,
        odataRequest: ODataRequest,
        _updateParent?: boolean
    ): Promise<any>;
    performPATCH(
        keyValues: KeyDefinitions,
        patchData: object,
        tenantId: string,
        odataRequest: ODataRequest,
        _updateParent?: boolean
    ): Promise<any>;
    performDELETE(
        keyValues: KeyDefinitions,
        tenantId: string,
        odataRequest: ODataRequest,
        _updateParent?: boolean
    ): Promise<void>;
    getParentEntityInterface(tenantId: string): Promise<FileBasedMockData | undefined>;
    getEntityInterface(entitySetName: string, tenantId: string): Promise<FileBasedMockData | undefined>;
    getMockData(tenantId: string): FileBasedMockData;
    isV4(): boolean;
}
export interface DataAccessInterface {
    isV4(): boolean;
    getNavigationPropertyKeys(
        data: any,
        navPropDetail: any,
        currentEntityType: EntityType,
        currentEntitySet: EntitySet | undefined,
        currentKeys: Record<string, string>,
        forCreate?: boolean
    ): Record<string, string>;
    getMockEntitySet(
        entityTypeName: string,
        generateMockData?: boolean,
        containedEntityType?: EntityType,
        containedData?: any
    ): Promise<EntitySetInterface>;
    getData(odataRequest: ODataRequest): Promise<any>;
    getDraftRoot(keyValues: KeyDefinitions, _tenantId: string, entitySetDefinition: EntitySet): any;
    getMetadata(): ODataMetadata;
    debug: boolean;
    fileLoader: IFileLoader;
    log: ILogger;
}

/**
 *
 */
export class ExecutionError extends Error {
    statusCode: number;
    messageData: object;
    isSAPMessage: boolean;
    isCustomError = true;
    headers: Record<string, string>;

    constructor(
        message: string,
        statusCode: number,
        messageData: any,
        isSAPMessage: boolean,
        headers: Record<string, string> = {}
    ) {
        super(message);
        this.statusCode = statusCode;
        this.messageData = messageData;
        this.isSAPMessage = isSAPMessage;
        this.headers = headers;
    }
}

export function _getDateTimeOffset(isV4: boolean) {
    const date = new Date();
    return isV4 ? date.toISOString() : '/Date(' + date.getTime() + '+0000)/';
}

const CHARS = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';

/**
 * Generate an ID of a given length.
 *
 * @param length
 * @returns the generated ID
 */
export function generateId(length: number) {
    let result = '';
    while (length--) {
        result += CHARS.charAt(Math.floor(Math.random() * CHARS.length));
    }
    return result;
}

export function uuidv4() {
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
        const r = (Math.random() * 16) | 0,
            v = c == 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
    });
}
