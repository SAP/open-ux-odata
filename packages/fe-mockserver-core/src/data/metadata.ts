import { parse } from '@sap-ux/edmx-parser';
import { convert } from '@sap-ux/annotation-converter';
import type {
    Action,
    ActionImport,
    ConvertedMetadata,
    EntitySet,
    EntityType,
    NavigationProperty,
    RawMetadata,
    Singleton
} from '@sap-ux/vocabularies-types';

type NameAndNav = {
    name: string;
    entitySet: EntitySet | Singleton;
    navigation: NavigationProperty;
    partner?: NavigationProperty;
};

/**
 *
 */
export class ODataMetadata {
    public static async parse(edmx: string, metadataUrl: string, ETag?: string): Promise<ODataMetadata> {
        const metadata: RawMetadata = parse(edmx);
        const converterMetadata: ConvertedMetadata = convert(metadata);
        return new ODataMetadata(converterMetadata, metadata, metadataUrl, edmx, ETag);
    }

    public getEdmx() {
        return this.edmx;
    }

    public getMetadataUrl(): string {
        return this.metadataUrl;
    }

    public readonly typeToEntityMapping: object = {};
    public readonly ETag: string | undefined;
    public readonly metadataUrl: string;
    protected readonly metadata: ConvertedMetadata;
    protected readonly parserMetadata: RawMetadata;

    private constructor(
        metadata: ConvertedMetadata,
        parserMetadata: RawMetadata,
        metadataUrl: string,
        private edmx: string,
        ETag?: string
    ) {
        this.metadata = metadata;
        this.parserMetadata = parserMetadata;
        this.metadataUrl = metadataUrl;
        this.ETag = ETag;
    }

    public getVersion(): string {
        return this.metadata.version === '1.0' ? '2.0' : '4.0';
    }

    public getEntitySet(entitySetName: string): EntitySet | undefined {
        return this.metadata.entitySets.find((entitySet) => entitySet.name === entitySetName);
    }

    public getSingleton(singletonName: string): EntitySet {
        return this.metadata.singletons.find((singleton) => singleton.name === singletonName) as any;
    }

    public getEntitySets(): EntitySet[] {
        return this.metadata.entitySets;
    }

    public getSingletons(): Singleton[] {
        return this.metadata.singletons || [];
    }

    public getEntityType(entityTypeName: string): EntityType | undefined {
        return this.metadata.entityTypes.find((entityType) => entityType.name === entityTypeName);
    }

    public getEntitySetByType(entityTypeName: string): EntitySet | undefined {
        return this.metadata.entitySets.find((entitySet) => entitySet.entityTypeName === entityTypeName);
    }

    public getActionByFQN(actionFQN: string): Action | undefined {
        return this.metadata.actions.find((action) => action.fullyQualifiedName === actionFQN);
    }

    public getActionImportByFQN(actionImportFQN: string): ActionImport | undefined {
        return this.metadata.actionImports.find((actionImport) => actionImport.fullyQualifiedName === actionImportFQN);
    }

    public getEntityContainerPath(): string {
        return this.parserMetadata.schema.entityContainer.fullyQualifiedName;
    }

    public isDraftEntity(entitySet: EntitySet): boolean {
        return (
            entitySet?.annotations?.Common?.DraftRoot !== undefined ||
            entitySet?.annotations?.Common?.DraftNode !== undefined
        );
    }

    public isStickyEntity(entitySet: EntitySet): boolean {
        return entitySet?.annotations?.Session?.StickySessionSupported !== undefined;
    }

    public isDraftRoot(entitySet: EntitySet): boolean {
        return entitySet?.annotations?.Common?.DraftRoot !== undefined;
    }

    public isDraftNode(entitySet: EntitySet | Singleton): boolean {
        return entitySet._type === 'EntitySet' && entitySet.annotations.Common?.DraftNode !== undefined;
    }

    public resolvePath(path: string): any {
        return this.metadata.resolvePath(path);
    }

    public findInDescendant(
        entitySet: EntitySet | Singleton,
        targetEntitySet: EntitySet,
        path: NameAndNav[],
        entitySetFilter: (entitySet: EntitySet | Singleton) => boolean = () => true
    ): boolean {
        if (entitySet === targetEntitySet) {
            // nothing to do - we are there already
            return true;
        }

        for (const [navPropBindingName, target] of Object.entries(entitySet.navigationPropertyBinding)) {
            if (!entitySetFilter(target)) {
                continue;
            }

            const navProp = entitySet.entityType.navigationProperties.find(
                (navigationProperty) => navigationProperty.name === navPropBindingName
            );

            if (path.some((segment) => segment.entitySet === target)) {
                // would add a cycle
                return false;
            } else if (navProp) {
                const partnerNavProp = navProp.targetType.navigationProperties.find(
                    (navigationProperty) => navigationProperty.name === navProp.partner
                );
                path.push({
                    name: navPropBindingName,
                    entitySet: target,
                    navigation: navProp,
                    partner: partnerNavProp
                });
            }

            if (target === targetEntitySet) {
                return true;
            } else {
                return this.findInDescendant(target, targetEntitySet, path, entitySetFilter);
            }
        }

        return false;
    }

    public resolveDraftRoot(entitySet: EntitySet) {
        if (this.isDraftEntity(entitySet)) {
            // entity set must be a draft node, otherwise there is no point in looking for its root
            const draftRoots = this.metadata.entitySets.filter(this.isDraftRoot);

            for (const draftRoot of draftRoots) {
                const path: NameAndNav[] = [];
                const found = this.findInDescendant(draftRoot, entitySet, path, this.isDraftNode);
                if (found) {
                    return {
                        found: true,
                        entitySet: draftRoot,
                        path: path
                    };
                }
            }
        }

        return { found: false, entitySet: undefined, path: [] };
    }

    public resolveAncestors(entitySet: EntitySet) {
        let found = false;
        let foundPath!: NameAndNav[];
        let rootEntitySet!: EntitySet;
        this.metadata.entitySets.forEach((et) => {
            if (!found) {
                const resolvePath: any[] = [];
                found = this.findInDescendant(et, entitySet, resolvePath);
                if (found) {
                    foundPath = resolvePath.concat();
                    rootEntitySet = et;
                }
            }
        });
        return {
            found,
            entitySet: rootEntitySet,
            path: foundPath
        };
    }

    public getParentEntitySetName(sourceEntitySet: EntitySet) {
        const { found, path, entitySet } = this.resolveAncestors(sourceEntitySet);
        if (found) {
            if (path.length === 1 && entitySet) {
                return entitySet.name;
            } else {
                return path[path.length - 1].entitySet.name;
            }
        }
    }

    public getETag() {
        return this.ETag;
    }

    public getKeys(dataLine: any, entityType: EntityType): Record<string, string | number | boolean> {
        const keys = entityType.keys;
        const keyValues: any = {};
        keys.forEach((keyProp) => {
            keyValues[keyProp.name] = dataLine[keyProp.name];
        });

        return keyValues;
    }
}
