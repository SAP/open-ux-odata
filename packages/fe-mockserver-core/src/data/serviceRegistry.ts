import type { DataAccessInterface } from './common';

/**
 * Registry for managing cross-service communication in the mockserver.
 * Allows services to access entity interfaces from other services in a controlled manner.
 */
export class ServiceRegistry {
    private readonly services: Map<string, DataAccessInterface> = new Map();

    /**
     * Register a service with its DataAccess instance
     * @param serviceName - The name/path of the service
     * @param dataAccess - The DataAccess instance for this service
     */
    public registerService(serviceName: string, dataAccess: DataAccessInterface): void {
        this.services.set(serviceName, dataAccess);
    }

    /**
     * Get a DataAccess instance for a specific service
     * @param serviceName - The name/path of the service
     * @returns The DataAccess instance or undefined if not found
     */
    public getService(serviceName: string): DataAccessInterface | undefined {
        return this.services.get(serviceName);
    }

    /**
     * Get all registered service names
     * @returns Array of service names
     */
    public getServiceNames(): string[] {
        return Array.from(this.services.keys());
    }
}
