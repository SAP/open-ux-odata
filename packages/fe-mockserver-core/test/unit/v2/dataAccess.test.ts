import { join } from 'path';
import { ODataMetadata } from '../../../src/data/metadata';
import { DataAccess } from '../../../src/data/dataAccess';
import type { ServiceConfig } from '../../../src';
import FileSystemLoader from '../../../src/plugins/fileSystemLoader';
import CDSMetadataProvider from '@sap-ux/fe-mockserver-plugin-cds';
import ODataRequest from '../../../src/request/odataRequest';

describe('Data Access', () => {
    let dataAccess!: DataAccess;
    let dataAccessNoMock!: DataAccess;
    let metadataNoMock!: ODataMetadata;
    let metadata!: ODataMetadata;
    const baseUrl = '/sap/fe/preview/Form';
    const fileLoader = new FileSystemLoader();
    const metadataProvider = new CDSMetadataProvider(fileLoader);
    beforeAll(async () => {
        const baseDir = join(__dirname, 'services', 'productSample');
        const baseDirNoMock = join(__dirname, 'services', 'productSampleNoMock');

        const edmx = await metadataProvider.loadMetadata(join(baseDir, 'metadata.xml'));
        const edmxNoMock = await metadataProvider.loadMetadata(join(baseDirNoMock, 'metadata.xml'));

        metadata = await ODataMetadata.parse(edmx, baseUrl + '/$metadata');
        metadataNoMock = await ODataMetadata.parse(edmxNoMock, baseUrl + '/$metadata');
        dataAccess = new DataAccess({ mockdataPath: baseDir } as ServiceConfig, metadata, fileLoader);
        dataAccessNoMock = new DataAccess(
            { mockdataPath: baseDirNoMock, generateMockData: true } as ServiceConfig,
            metadata,
            fileLoader
        );
    });
    test('v2metadata - it can GET data for an entity', async () => {
        let odataRequest = new ODataRequest(
            {
                method: 'GET',
                url: '/SEPMRA_C_PD_Product?$skip=0&$top=20&$orderby=to_ProductTextInOriginalLang/Name%20asc&$filter=IsActiveEntity%20eq%20false%20or%20SiblingEntity/IsActiveEntity%20eq%20null&$select=ProductCategory%2cProductPictureURL%2cProductForEdit%2cto_ProductTextInOriginalLang%2fName%2cProduct%2cDraftUUID%2cIsActiveEntity%2cHasDraftEntity%2cHasActiveEntity%2cCopy_ac%2cDraftAdministrativeData&$expand=to_ProductTextInOriginalLang%2cDraftAdministrativeData&$inlinecount=allpages'
            },
            dataAccess
        );
        let productData = await dataAccess.getData(odataRequest);
        expect(productData.length).toEqual(4);
        expect(odataRequest.dataCount).toEqual(4);
        odataRequest = new ODataRequest({ method: 'GET', url: '/SEPMRA_C_PD_Product?$skip=0&$top=3' }, dataAccess);
        productData = await dataAccess.getData(odataRequest);
        expect(productData.length).toEqual(3);
        expect(odataRequest.dataCount).toEqual(4);
        expect(productData[0].Supplier).toEqual('Acme DE');
        expect(productData[0].Price).toEqual(3);

        productData = await dataAccess.getData(
            new ODataRequest({ method: 'GET', url: '/SEPMRA_C_PD_Product?$skip=1&$top=3' }, dataAccess)
        );
        expect(productData.length).toEqual(3);
        expect(productData[0].Supplier).toEqual('Acme DE');
        expect(productData[0].Price).toEqual(1);
        productData = await dataAccess.getData(
            new ODataRequest({ method: 'GET', url: '/SEPMRA_C_PD_Product?$select=Supplier' }, dataAccess)
        );
        expect(productData[0].Supplier).toEqual('Acme DE');
        expect(productData[0].Name).toBeUndefined();
        productData = await dataAccess.getData(
            new ODataRequest({ method: 'GET', url: "/SEPMRA_C_PD_Product?$filter=Supplier eq 'Acme DE'" }, dataAccess)
        );
        expect(productData.length).toEqual(2);
        expect(productData[0].Supplier).toEqual('Acme DE');
        expect(productData[0].Name).toEqual('Acme Boomerang');
        productData = await dataAccess.getData(
            new ODataRequest(
                {
                    method: 'GET',
                    url: "/SEPMRA_C_PD_Product?$filter=StartingSaleDate gt datetime'2022-10-07T06:52:24.189'"
                },
                dataAccess
            )
        );
        expect(productData.length).toEqual(2);
        expect(productData[0].Supplier).toEqual('Acme DE');
        expect(productData[0].Name).toEqual('Acme TNT');
        expect(productData[1].Name).toEqual('Acme Extra Comfy Safe');
        productData = await dataAccess.getData(
            new ODataRequest({ method: 'GET', url: '/SEPMRA_C_PD_Product?$orderby=Price,Supplier' }, dataAccess)
        );
        expect(productData.length).toEqual(4);
        expect(productData[0].Name).toEqual('Acme TNT');
        expect(productData[1].Name).toEqual('Acme Trap v2');
        expect(productData[2].Name).toEqual('Acme Boomerang');
        expect(productData[3].Name).toEqual('Acme Extra Comfy Safe');
    });
    test('v2dataAccess - it can GET data for an entity when there is no mock data', async () => {
        const odataRequest = new ODataRequest(
            {
                method: 'GET',
                url: '/SEPMRA_C_PD_Product?$skip=0&$top=20&$orderby=to_ProductTextInOriginalLang/Name%20asc&$filter=IsActiveEntity%20eq%20false%20or%20SiblingEntity/IsActiveEntity%20eq%20null&$select=ProductPictureURL%2cProductForEdit%2cProductForEdit_fc%2cto_ProductTextInOriginalLang%2fName%2cto_ProductCategory%2fMainProductCategory%2cProductCategory%2cto_Supplier%2fCompanyName%2cto_Supplier%2cto_ProductStock%2fStockAvailability%2cto_ProductStock%2fto_StockAvailability%2fStockAvailability_Text%2cto_CollaborativeReview%2fAverageRatingValue%2cPrice%2cCurrency%2cProduct%2cDraftUUID%2cIsActiveEntity%2cHasDraftEntity%2cHasActiveEntity%2cCopy_ac%2cDraftAdministrativeData&$expand=to_ProductCategory%2cto_Supplier%2cto_ProductStock%2cto_ProductStock%2fto_StockAvailability%2cto_ProductTextInOriginalLang%2cto_CollaborativeReview%2cDraftAdministrativeData&$inlinecount=allpages'
            },
            dataAccess
        );
        const productData = await dataAccessNoMock.getData(odataRequest);
        expect(productData.length).toEqual(20);
        expect(odataRequest.dataCount).toEqual(150);
    });
    test('v2metadata - it can GET data for an entity', async () => {
        const odataRequest = new ODataRequest(
            {
                method: 'GET',
                url: '/SEPMRA_C_PD_Product?$skip=0&$top=20&$orderby=to_ProductTextInOriginalLang/Name%20asc&$filter=IsActiveEntity%20eq%20false%20or%20SiblingEntity/IsActiveEntity%20eq%20null&$select=ProductCategory%2cProductPictureURL%2cProductForEdit%2cto_ProductTextInOriginalLang%2fName%2cProduct%2cDraftUUID%2cIsActiveEntity%2cHasDraftEntity%2cHasActiveEntity%2cCopy_ac%2cDraftAdministrativeData&$expand=to_ProductTextInOriginalLang%2cDraftAdministrativeData&$inlinecount=allpages'
            },
            dataAccess
        );
        await odataRequest.handleRequest();
        let responseData = odataRequest.getResponseData() || '';
        // remove data
        responseData = responseData.replace(/\/Date\([^)]+\)/g, '/Date()');
        expect(responseData).toMatchSnapshot();
        expect(odataRequest.responseHeaders).toMatchSnapshot();
        const odataRequestSolo = new ODataRequest(
            {
                method: 'GET',
                url: "/SEPMRA_C_PD_Product(Product='Acme_Boomerang',DraftUUID='',IsActiveEntity=true)"
            },
            dataAccess
        );
        await odataRequestSolo.handleRequest();
        let responseDataSolo = odataRequestSolo.getResponseData() || '';
        responseDataSolo = responseDataSolo.replace(/\/Date\([^)]+\)/g, '/Date()');
        expect(responseDataSolo).toMatchSnapshot();
        expect(odataRequestSolo.responseHeaders).toMatchSnapshot();
    });
    test('v2metadata - it can DELETE data for an entity', async () => {
        const odataRequestDelete = new ODataRequest(
            {
                method: 'DELETE',
                url: "/SEPMRA_C_PD_Product(Product='Acme_Boomerang',DraftUUID='',IsActiveEntity=true)"
            },
            dataAccess
        );
        await odataRequestDelete.handleRequest();
        const responseDataDelete = odataRequestDelete.getResponseData();
        expect(responseDataDelete).toMatchSnapshot();
        expect(odataRequestDelete.responseHeaders).toMatchSnapshot();
    });
});
