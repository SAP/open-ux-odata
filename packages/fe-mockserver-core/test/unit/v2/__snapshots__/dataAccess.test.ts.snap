// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Access v2metadata - it can DELETE data for an entity 1`] = `undefined`;

exports[`Data Access v2metadata - it can DELETE data for an entity 2`] = `
{
  "cache-control": "no-store, no-cache",
  "content-type": "application/json",
  "dataserviceversion": "2.0",
  "sap-tenantid": "tenant-default",
}
`;

exports[`Data Access v2metadata - it can GET data for an entity 1`] = `"{"d":{"__count":4,"results":[{"Product":"Acme_Boomerang","DraftUUID":"","IsActiveEntity":true,"HasActiveEntity":true,"HasDraftEntity":false,"Copy_ac":false,"ProductForEdit":"","ProductCategory":"","ProductPictureURL":"","to_ProductTextInOriginalLang":{"Preparation_ac":false,"Validation_ac":false,"Description_fc":0,"LanguageForEdit_fc":0,"Delete_mc":false,"Product":"","Language":"","LanguageForEdit":"","Name":"","Description":"","DraftUUID":"","DraftEntityCreationDateTime":"/Date()/","DraftEntityLastChangeDateTime":"/Date()/","ParentDraftUUID":"","HasActiveEntity":false,"HasDraftEntity":false,"IsActiveEntity":false,"__metadata":{"id":"undefined/SEPMRA_C_PD_ProductText(Product='',Language='',DraftUUID='',IsActiveEntity='false')","uri":"undefined/SEPMRA_C_PD_ProductText(Product='',Language='',DraftUUID='',IsActiveEntity='false')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductTextType"}},"DraftAdministrativeData":{"DraftUUID":"","DraftEntityType":"","CreationDateTime":"/Date()/","CreatedByUser":"","LastChangeDateTime":"/Date()/","LastChangedByUser":"","DraftAccessType":"","ProcessingStartDateTime":"/Date()/","InProcessByUser":"","DraftIsKeptByUser":false,"EnqueueStartDateTime":"/Date()/","DraftIsCreatedByMe":false,"DraftIsLastChangedByMe":false,"DraftIsProcessedByMe":false,"CreatedByUserDescription":"","LastChangedByUserDescription":"","InProcessByUserDescription":"","__metadata":{"id":"undefined/I_DraftAdministrativeData()","uri":"undefined/I_DraftAdministrativeData()","type":"SEPMRA_PROD_MAN.I_DraftAdministrativeDataType"}},"__metadata":{"id":"undefined/SEPMRA_C_PD_Product(Product='Acme_Boomerang',DraftUUID='',IsActiveEntity='true')","uri":"undefined/SEPMRA_C_PD_Product(Product='Acme_Boomerang',DraftUUID='',IsActiveEntity='true')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductType"}},{"Product":"Acme_TNT","DraftUUID":"","IsActiveEntity":true,"HasActiveEntity":true,"HasDraftEntity":false,"Copy_ac":false,"ProductForEdit":"","ProductCategory":"","ProductPictureURL":"","to_ProductTextInOriginalLang":{"Preparation_ac":false,"Validation_ac":false,"Description_fc":0,"LanguageForEdit_fc":0,"Delete_mc":false,"Product":"","Language":"","LanguageForEdit":"","Name":"","Description":"","DraftUUID":"","DraftEntityCreationDateTime":"/Date()/","DraftEntityLastChangeDateTime":"/Date()/","ParentDraftUUID":"","HasActiveEntity":false,"HasDraftEntity":false,"IsActiveEntity":false,"__metadata":{"id":"undefined/SEPMRA_C_PD_ProductText(Product='',Language='',DraftUUID='',IsActiveEntity='false')","uri":"undefined/SEPMRA_C_PD_ProductText(Product='',Language='',DraftUUID='',IsActiveEntity='false')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductTextType"}},"DraftAdministrativeData":{"DraftUUID":"","DraftEntityType":"","CreationDateTime":"/Date()/","CreatedByUser":"","LastChangeDateTime":"/Date()/","LastChangedByUser":"","DraftAccessType":"","ProcessingStartDateTime":"/Date()/","InProcessByUser":"","DraftIsKeptByUser":false,"EnqueueStartDateTime":"/Date()/","DraftIsCreatedByMe":false,"DraftIsLastChangedByMe":false,"DraftIsProcessedByMe":false,"CreatedByUserDescription":"","LastChangedByUserDescription":"","InProcessByUserDescription":"","__metadata":{"id":"undefined/I_DraftAdministrativeData()","uri":"undefined/I_DraftAdministrativeData()","type":"SEPMRA_PROD_MAN.I_DraftAdministrativeDataType"}},"__metadata":{"id":"undefined/SEPMRA_C_PD_Product(Product='Acme_TNT',DraftUUID='',IsActiveEntity='true')","uri":"undefined/SEPMRA_C_PD_Product(Product='Acme_TNT',DraftUUID='',IsActiveEntity='true')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductType"}},{"Product":"Acme_Trap","DraftUUID":"","IsActiveEntity":true,"HasActiveEntity":true,"HasDraftEntity":false,"Copy_ac":false,"ProductForEdit":"","ProductCategory":"","ProductPictureURL":"","to_ProductTextInOriginalLang":{"Preparation_ac":false,"Validation_ac":false,"Description_fc":0,"LanguageForEdit_fc":0,"Delete_mc":false,"Product":"","Language":"","LanguageForEdit":"","Name":"","Description":"","DraftUUID":"","DraftEntityCreationDateTime":"/Date()/","DraftEntityLastChangeDateTime":"/Date()/","ParentDraftUUID":"","HasActiveEntity":false,"HasDraftEntity":false,"IsActiveEntity":false,"__metadata":{"id":"undefined/SEPMRA_C_PD_ProductText(Product='',Language='',DraftUUID='',IsActiveEntity='false')","uri":"undefined/SEPMRA_C_PD_ProductText(Product='',Language='',DraftUUID='',IsActiveEntity='false')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductTextType"}},"DraftAdministrativeData":{"DraftUUID":"","DraftEntityType":"","CreationDateTime":"/Date()/","CreatedByUser":"","LastChangeDateTime":"/Date()/","LastChangedByUser":"","DraftAccessType":"","ProcessingStartDateTime":"/Date()/","InProcessByUser":"","DraftIsKeptByUser":false,"EnqueueStartDateTime":"/Date()/","DraftIsCreatedByMe":false,"DraftIsLastChangedByMe":false,"DraftIsProcessedByMe":false,"CreatedByUserDescription":"","LastChangedByUserDescription":"","InProcessByUserDescription":"","__metadata":{"id":"undefined/I_DraftAdministrativeData()","uri":"undefined/I_DraftAdministrativeData()","type":"SEPMRA_PROD_MAN.I_DraftAdministrativeDataType"}},"__metadata":{"id":"undefined/SEPMRA_C_PD_Product(Product='Acme_Trap',DraftUUID='',IsActiveEntity='true')","uri":"undefined/SEPMRA_C_PD_Product(Product='Acme_Trap',DraftUUID='',IsActiveEntity='true')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductType"}},{"Product":"Acme_Safe","DraftUUID":"","IsActiveEntity":true,"HasActiveEntity":true,"HasDraftEntity":false,"Copy_ac":false,"ProductForEdit":"","ProductCategory":"","ProductPictureURL":"","to_ProductTextInOriginalLang":{"Preparation_ac":false,"Validation_ac":false,"Description_fc":0,"LanguageForEdit_fc":0,"Delete_mc":false,"Product":"","Language":"","LanguageForEdit":"","Name":"","Description":"","DraftUUID":"","DraftEntityCreationDateTime":"/Date()/","DraftEntityLastChangeDateTime":"/Date()/","ParentDraftUUID":"","HasActiveEntity":false,"HasDraftEntity":false,"IsActiveEntity":false,"__metadata":{"id":"undefined/SEPMRA_C_PD_ProductText(Product='',Language='',DraftUUID='',IsActiveEntity='false')","uri":"undefined/SEPMRA_C_PD_ProductText(Product='',Language='',DraftUUID='',IsActiveEntity='false')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductTextType"}},"DraftAdministrativeData":{"DraftUUID":"","DraftEntityType":"","CreationDateTime":"/Date()/","CreatedByUser":"","LastChangeDateTime":"/Date()/","LastChangedByUser":"","DraftAccessType":"","ProcessingStartDateTime":"/Date()/","InProcessByUser":"","DraftIsKeptByUser":false,"EnqueueStartDateTime":"/Date()/","DraftIsCreatedByMe":false,"DraftIsLastChangedByMe":false,"DraftIsProcessedByMe":false,"CreatedByUserDescription":"","LastChangedByUserDescription":"","InProcessByUserDescription":"","__metadata":{"id":"undefined/I_DraftAdministrativeData()","uri":"undefined/I_DraftAdministrativeData()","type":"SEPMRA_PROD_MAN.I_DraftAdministrativeDataType"}},"__metadata":{"id":"undefined/SEPMRA_C_PD_Product(Product='Acme_Safe',DraftUUID='',IsActiveEntity='true')","uri":"undefined/SEPMRA_C_PD_Product(Product='Acme_Safe',DraftUUID='',IsActiveEntity='true')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductType"}}]}}"`;

exports[`Data Access v2metadata - it can GET data for an entity 2`] = `
{
  "cache-control": "no-store, no-cache",
  "content-type": "application/json",
  "dataserviceversion": "2.0",
  "sap-tenantid": "tenant-default",
}
`;

exports[`Data Access v2metadata - it can GET data for an entity 3`] = `"{"d":{"Product":"Acme_Boomerang","DraftUUID":"","Price":3,"Name":"Acme Boomerang","Supplier":"Acme DE","StartingSaleDate":"2018-01-01T00:00:00","IsActiveEntity":true,"HasActiveEntity":true,"HasDraftEntity":false,"Activation_ac":false,"Copy_ac":false,"Copy_new_supplier_ac":false,"Create_review_post_ac":false,"Delete_ext_ac":false,"Edit_ac":false,"Favorites_add_ac":false,"Favorites_remove_ac":false,"Favorites_toggle_ac":false,"Mycart_add_ac":false,"Preparation_ac":false,"Shortage_list_ac":false,"Validation_ac":false,"Description_fc":0,"ProductForEdit_fc":0,"ProductForEdit":"","ProductCategory":"","Currency":"","Height":0,"Width":0,"Depth":0,"DimensionUnit":"","ProductPictureURL":"","ProductBaseUnit":"","Weight":0,"WeightUnit":"","OriginalLanguage":"","Description":"","DraftEntityCreationDateTime":"/Date()/","DraftEntityLastChangeDateTime":"/Date()/","__metadata":{"id":"undefined/SEPMRA_C_PD_Product(Product='Acme_Boomerang',DraftUUID='',IsActiveEntity='true')","uri":"undefined/SEPMRA_C_PD_Product(Product='Acme_Boomerang',DraftUUID='',IsActiveEntity='true')","type":"SEPMRA_PROD_MAN.SEPMRA_C_PD_ProductType"}}}"`;

exports[`Data Access v2metadata - it can GET data for an entity 4`] = `
{
  "cache-control": "no-store, no-cache",
  "content-type": "application/json",
  "dataserviceversion": "2.0",
  "sap-tenantid": "tenant-default",
}
`;
